### PR TITLE
[2.9] Allow bundle to update existing offers

### PR DESCRIPTION
**Pull request by achilleasa**
This is a forward port of #12608 to 2.9. Due to the number of changes between 2.8 and 2.9 (notably: bundlechanges got bumped from v1 to v4), the commits where cherry-picked and some minor conflicts have been resolved.

## QA steps

Follow the QA steps from #12608 

## Bug reference

https://bugs.launchpad.net/juju/+bug/1894004

### CODE REVIEW

File: cmd/juju/application/deployer/bundlehandler.go [hunk 1]
```diff
@@ -1439,8 +1437,196 @@ func resolve(placeholder string, results map[string]string) string {
...
+		return reflect.DeepEqual(ac, bc)
+	}
+
+	return mod, nil
+}
+
+// applicationConfigValue returns the value if it is not a default value.
+// If the value is a default value, nil is returned.
+// If there was issue determining the type or value, an error is returned.
+func applicationConfigValue(key string, valueMap interface{}) (interface{}, error) {
```
Comment (achilleasa) [inline cmd/juju/application/deployer/bundlehandler.go:1609]: This got changed due to the merge conflict. Not sure if this got added in 2.9 or it is something that was removed with the deploy process overhaul.

Comment (hmlanigan) [reply cmd/juju/application/deployer/bundlehandler.go:1609]: applicationConfigValue is now part of the bundle package: https://github.com/juju/juju/blob/f76391aee8f3ce5eb07eaddd8397464f5d48f3e8/cmd/juju/application/bundle/bundle.go#L190

File: cmd/juju/application/deployer/bundlehandler.go [hunk 2]
```diff
@@ -1439,8 +1437,196 @@ func resolve(placeholder string, results map[string]string) string {
...
+// ModelExtractor provides everything we need to build a
+// bundlechanges.Model from a model API connection.
+type ModelExtractor interface {
+	GetAnnotations(tags []string) ([]params.AnnotationsGetResult, error)
+	GetConstraints(applications ...string) ([]constraints.Value, error)
+	GetConfig(branchName string, applications ...string) ([]map[string]interface{}, error)
+	Sequences() (map[string]int, error)
+}
+
+func buildModelRepresentation(
```
Comment (hmlanigan) [inline cmd/juju/application/deployer/bundlehandler.go:1449]: BuildModelRepresentation and the ModelExtactor interface were moved to https://github.com/juju/juju/blob/f76391aee8f3ce5eb07eaddd8397464f5d48f3e8/cmd/juju/application/bundle/bundle.go#L28 in 2.9. To be used for deploy and bundlediff.  The small change from the PR can be made there.

I do not believe that the content was changed, just the location.

**Review by hmlanigan — CHANGES_REQUESTED** [review]
The 2.8 commit 82524891a82e3f0920fccefc0a8755c8373040df can be merged against cmd/juju/applications/bundle/bundle.go in 2.9.

**Comment by achilleasa** [comment]
@hmlanigan I updated the PR so everything should hopefully have landed at the right place now. Can you take another look?

**Review by hmlanigan — APPROVED** [review]
LGTM, the cmd bundle.go changes are inline with the refactored deploy/refresh code in 2.9.

**Comment by achilleasa** [comment]
$$merge$$

**Pull request merged**